src/Conversion/KrnlToLLVM/KrnlCall.cpp

```diff
@@ -0,0 +1,214 @@
+//
+// This file lowers the KrnlCallOp operator.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/LLVMCommon/Pattern.h"
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+
+#include "src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.hpp"
+#include "src/Dialect/Krnl/DialectBuilder.hpp"
+#include "src/Dialect/Krnl/KrnlHelper.hpp"
+#include "src/Dialect/Krnl/KrnlOps.hpp"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "krnl_to_llvm"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+namespace krnl {
+
+class KrnlCallOpLowering : public ConversionPattern {
+public:
+  explicit KrnlCallOpLowering(
+      TypeConverter &typeConverter, MLIRContext *context)
+      : ConversionPattern(
+            typeConverter, KrnlCallOp::getOperationName(), 1, context) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    KrnlCallOpAdaptor krnlCallAdaptor(operands);
+    auto loc = op->getLoc();
+    KrnlCallOp krnlCallOp = llvm::dyn_cast<KrnlCallOp>(op);
+
+    // Get a symbol reference to the function, inserting it if necessary.
+    ModuleOp module = op->getParentOfType<ModuleOp>();
+    llvm::SmallVector<Type, 4> parameterTypeList;
+    llvm::SmallVector<Value, 4> parameterList;
+    handleOneParameter(rewriter, op, krnlCallAdaptor.result(),
+        krnlCallOp.result(), parameterTypeList, parameterList);
+
+    // Some type of operands has been converted.
+    // It is better to check the type of original operands.
+    // Thus, the two kinds of operands are used together.
+    auto itConverted = krnlCallAdaptor.parameters().begin();
+    auto itOriginal = krnlCallOp.parameters().begin();
+    for (; itConverted != krnlCallAdaptor.parameters().end();
+         itConverted++, itOriginal++) {
+      handleOneParameter(rewriter, op, *itConverted, *itOriginal,
+          parameterTypeList, parameterList);
+    }
+
+    // Handle the Attributes
+    for (auto namedAttr : op->getAttrs()) {
+      // Avoid the funcName() Attribute
+      if (namedAttr.getName().getValue().equals("funcName"))
+        continue;
+      handleOneAttribute(rewriter, getTypeConverter(), op, namedAttr.getValue(),
+          parameterTypeList, parameterList);
+    }
+
+    auto callRef = getOrInsertCall(
+        rewriter, module, krnlCallOp.funcName(), parameterTypeList);
+    rewriter.create<func::CallOp>(
+        loc, callRef, ArrayRef<Type>({}), parameterList);
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+private:
+  static void handleOneParameter(PatternRewriter &rewriter, Operation *op,
+      Value parameter, Value original,
+      llvm::SmallVector<Type, 4> &parameterTypeList,
+      llvm::SmallVector<Value, 4> &parameterList) {
+    auto *context = op->getContext();
```

Comment:
LLVM code guidelines indicate that if the type of the RHS expression is not immediately clear from the context one should use the actual type rather than `auto`.